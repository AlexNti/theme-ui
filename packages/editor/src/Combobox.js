/** @jsx jsx */
import { jsx } from 'theme-ui'
import { useState, useRef, useEffect } from 'react'
import Label from './Label'
import Input from './Input'

const keys = {
  38: 'up',
  40: 'down',
  27: 'escape',
  13: 'return',
}

const Chevron = props => (
  <svg
    viewBox="0 0 16 16"
    width="16"
    height="16"
    sx={{
      pointerEvents: 'none',
    }}>
    <path
      stroke="currentcolor"
      strokeWidth="2"
      fill="none"
      d="M14 6 L8 12 L2 6"
    />
  </svg>
)

export default ({
  type = 'text',
  name,
  label,
  value,
  onChange,
  options = [],
  ...props
}) => {
  const [open, setOpen] = useState(false)
  const [index, setIndex] = useState(-1)
  const root = useRef(null)
  const input = useRef(null)

  useEffect(() => {
    const handleOutsideClick = e => {
      if (root.current && root.current.contains(e.target)) return
      setOpen(false)
    }
    document.addEventListener('click', handleOutsideClick)
    return () => {
      document.removeEventListener('click', handleOutsideClick)
    }
  }, [root.current])

  useEffect(() => {
    const i = options.indexOf(value)
    if (i < 0) return
    setIndex(i)
  }, [])

  const popup = name + '-popup'

  const handleKeyDown = e => {
    if (e.shiftKey || e.metaKey || e.ctrlKey || e.altKey) return
    switch (keys[e.keyCode]) {
      case 'up':
        if (!open) setOpen(true)
        if (index < 1) setIndex(options.length - 1)
        else setIndex(index - 1)
        break
      case 'down':
        if (!open) setOpen(true)
        setIndex((index + 1) % options.length)
        break
      case 'return':
        const val = options[index]
        if (val) onChange(val)
        setOpen(false)
        break
      case 'escape':
        setOpen(false)
        setIndex(-1)
        break
    }
  }

  const handleChange = e => {
    onChange(e.target.value)
  }

  const toggleOpen = e => {
    setOpen(!open)
    if (input.current) input.current.focus()
  }
  const handleItemClick = i => e => {
    const val = options[i]
    if (val) onChange(val)
    setOpen(false)
  }

  return (
    <div
      ref={root}
      style={{
        position: open ? 'relative' : 'static',
      }}
      sx={{
        zIndex: 2,
      }}>
      <Label htmlFor={name}>{label || name}</Label>
      <div
        sx={{
          display: 'flex',
        }}>
        <Input
          {...props}
          ref={input}
          role="combobox"
          type={type}
          id={name}
          name={name}
          value={value}
          onChange={handleChange}
          onKeyDown={handleKeyDown}
          aria-autocomplete="none"
          aria-haspopup="true"
          aria-owns={popup}
          aria-expanded={open}
          aria-activedescendant={name + index}
        />
        <button
          tabindex="-1"
          aria-label={open ? 'Close' : 'Open'}
          onClick={toggleOpen}
          sx={{
            appearance: 'none',
            width: 32,
            height: 32,
            color: 'inherit',
            bg: 'transparent',
            border: 0,
            ml: -32,
          }}>
          <Chevron />
        </button>
      </div>
      <ul
        id={popup}
        role="listbox"
        aria-label={name}
        tabindex="-1"
        style={{
          visibility: open ? 'visible' : 'hidden',
          position: open ? 'absolute' : 'static',
        }}
        sx={{
          left: 0,
          right: 0,
          top: '100%',
          listStyle: 'none',
          p: 0,
          m: 0,
          maxHeight: 512,
          overflowY: 'auto',
          bg: 'background',
          border: '1px solid',
          borderTop: 0,
          borderColor: 'gray',
        }}>
        {open &&
          options.map((option, i) => (
            <li
              key={option}
              role="option"
              aria-selected={i === index}
              onClick={handleItemClick(i)}
              sx={{
                p: 1,
                userSelect: 'none',
                '&[aria-selected=true],:hover': {
                  bg: 'highlight',
                },
              }}>
              {option}
            </li>
          ))}
      </ul>
    </div>
  )
}
