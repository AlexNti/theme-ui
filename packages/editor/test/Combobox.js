import React from 'react'
import renderer from 'react-test-renderer'
import { render, fireEvent, waitForElement, cleanup } from '@testing-library/react'
import Combobox from '../src/Combobox'

afterEach(cleanup)

test('renders', () => {
  const json = renderer.create(
    <Combobox />
  ).toJSON()
  expect(json).toMatchSnapshot()
})

test('clicking chevron button shows menu', async () => {
  const tree = render(
    <Combobox
      label='Beep'
      name='beep'
      options={[
        'beep',
        'boop',
      ]}
    />
  )
  const button = await tree.getByRole('button')
  fireEvent.click(button)
  const options = await waitForElement(() => tree.findAllByRole('option'))
  expect(options.length).toBe(2)
})

test.todo('clicking item updates value')
test.todo('down arrow moves selection down')
test.todo('up arrow moves selection up')
test.todo('return key selects item')
test.todo('escape key closes listbox')
test.todo('blur closes listbox')
