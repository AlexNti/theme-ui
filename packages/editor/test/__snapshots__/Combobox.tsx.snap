// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
.emotion-6 {
  z-index: 2;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-3 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 32px;
  height: 32px;
  color: inherit;
  background-color: transparent;
  border: 0;
  margin-left: -32px;
}

.emotion-2 {
  pointer-events: none;
}

.emotion-5 {
  left: 0;
  right: 0;
  top: 100%;
  list-style: none;
  padding: 0;
  margin: 0;
  max-height: 512px;
  overflow-y: auto;
  background-color: white;
  border: 1px solid;
  border-top: 0;
  border-color: lightgray;
}

<div
  className="emotion-6"
  style={
    Object {
      "position": "static",
    }
  }
>
  <label
    __css={
      Object {
        "display": "flex",
        "width": "100%",
      }
    }
    __themeKey="forms"
    className="emotion-0"
    htmlFor="beep"
  >
    beep
  </label>
  <div
    className="emotion-4"
  >
    <input
      __css={
        Object {
          "appearance": "none",
          "bg": "transparent",
          "border": "1px solid",
          "borderRadius": 4,
          "color": "inherit",
          "display": "block",
          "fontSize": "inherit",
          "lineHeight": "inherit",
          "p": 2,
          "width": "100%",
        }
      }
      __themeKey="forms"
      aria-activedescendant="beep-1"
      aria-autocomplete="none"
      aria-expanded={false}
      aria-haspopup="true"
      aria-owns="beep-popup"
      className="emotion-0"
      id="beep"
      name="beep"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      role="combobox"
      type="text"
      value="beep"
    />
    <button
      aria-label="Open"
      className="emotion-3"
      onClick={[Function]}
      tabIndex={-1}
    >
      <svg
        className="emotion-2"
        height="12"
        viewBox="0 0 16 16"
        width="12"
      >
        <path
          d="M14 6 L8 12 L2 6"
          fill="none"
          stroke="currentcolor"
          strokeWidth="2"
        />
      </svg>
    </button>
  </div>
  <ul
    aria-label="beep"
    className="emotion-5"
    id="beep-popup"
    role="listbox"
    style={
      Object {
        "position": "static",
        "visibility": "hidden",
      }
    }
    tabIndex={-1}
  />
</div>
`;
